v2/router.go

```diff
@@ -48,9 +48,57 @@ type Procedure struct {
 	Signature string
 }
 
+// EncodingProcedure specifies a single encoding-level handler and is mapped to a TransportProcedure by the router
+type EncodingProcedure struct {
+	// Name of the procedure.
+	Name string
+
+	// Service or empty to use the default service name.
+	Service string
+
+	// HandlerSpec specifying which handler and rpc type.
+	HandlerSpec EncodingHandlerSpec
+
+	// Encoding of the handler.
+	// (if present).
+	Encoding Encoding
+
+	// Human-readable signature of the handler.
+	Signature string
+
+	// Codec to assist mapping the encoding-level handler to transport-level handler
+	Codec InboundCodec
+}
+
+func (p EncodingProcedure) TransportProcedureFunc(c context.Context, r *Request, b *Buffer) (*Response, *Buffer, error) {
```

Comment:
Clever reuse of the EncodingProcedure struct.

On the one hand, you can avoid having to wrap with `yarpc.TransportProcedureFunc(f)` by making the `*EncodingProcedure` type implement `Handle(Context, Request, Buffer)(Response,Buffer, Error)` directly.

However, we would like to avoid making this method public API. We are generally cautious about revealing implementation details in ways that can’t be changed after release, and removing this function would be a breaking change.

You can make this private to the `yarpcrouter` package by creating a type that privately contains a an `EncodingProcedure` and implements `Handle(Context,Request,Buffer)(Response,Buffer,Error)` and use a type assertion to verify that the object implements `UnaryTransportHandler`. Details below.